### ISSUE TITLE
`AnyMut` allows aliasing

### ISSUE THREAD

**Issue by Kijewski** [stijnfrishert/sashay#5]
Unless I'm misunderstanding something, this example should not compile:

```rust
let mut data = 'z';
let mut any = AnyMut::erase(&mut data);
let a = any.downcast_mut::<char>().unwrap();
let b = any.downcast_mut::<char>().unwrap(); // this line should not compile
let c = any.downcast_mut::<char>().unwrap();
*a = 'a';
*b = 'b';
*c = 'c';
```

This can be fixed by using explicit lifetimes:

```rust
pub fn downcast_mut<'b, U: 'static>(&'b mut self) -> Option<&'b mut U> where 'a: 'b;
pub fn downcast_ref<'b, T: 'static>(&'b self) -> Option<&'b T> where 'a: 'b;
```

**Comment by stijnfrishert** [comment]
Wow, hadn't expected anyone to actually come here 9h within me publishing, haha. How did you find this crate?

Thanks so much for the feedback. You are completely right, it seems. I believe I had that lifetime bound in one version, but erroneous reasoning made me remove it.

I'll update this asap.

**Comment by stijnfrishert** [comment]
Your example did indeed not pass `cargo miri test`. Added the lifetime bounds, checked again, and after everything passed I merged this into `main`.

Does this look alright now, @Kijewski? If so, I'll release this as `0.2.0` and yank the earlier versions.

**Comment by Kijewski** [comment]
Yes, now it looks good to me! (I did not review everything properly, though.)

I was visiting crates.io to get to the docs of another crate, and though "lifetime erased references" sounded interesting, so I looked at the code. :)

**Comment by stijnfrishert** [comment]
Alright, thank you again for your help ^_^

**Comment by stijnfrishert** [comment]
Just released this as 0.2.0.
